### PR TITLE
[stdlib] Fix reference type mutation issues in RangeReplaceableCollectionType

### PR DESCRIPTION
**Pull request by nielsandriesse**
Successfully passed `build-script -t`

The overloads for `CollectionType` appear to be redundant, because `underestimateCount()` returns `count` for a `CollectionType`. Is there any reason to include them?

### CODE REVIEW

File: stdlib/public/core/RangeReplaceableCollectionType.swift [hunk 1]
```diff
@@ -446,10 +446,10 @@ public func +<
     S : SequenceType
     where S.Generator.Element == C.Generator.Element
 >(lhs: C, rhs: S) -> C {
-  var lhs = lhs
-  // FIXME: what if lhs is a reference type?  This will mutate it.
-  lhs.appendContentsOf(rhs)
-  return lhs
+  var result = lhs is AnyObject ? C(lhs) : lhs
```
Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: A thing does not need to be AnyObject to have reference semantics -- for example, a struct wrapping a class also has reference semantics.

File: stdlib/public/core/RangeReplaceableCollectionType.swift [hunk 2]
```diff
@@ -471,11 +471,23 @@ public func +<
...
+  result.reserveCapacity(lhs.count + numericCast(rhs.count))
+  result.appendContentsOf(rhs)
+  return result
+}
+
+@warn_unused_result
+public func +<
```
Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: Why is this extra overload needed?  Why CollectionType and not SequenceType?

File: stdlib/public/core/RangeReplaceableCollectionType.swift [hunk 3]
```diff
@@ -459,7 +459,7 @@ public func +<
     where S.Generator.Element == C.Generator.Element
 >(lhs: S, rhs: C) -> C {
   var result = C()
-  result.reserveCapacity(rhs.count + numericCast(rhs.underestimateCount()))
+  result.reserveCapacity(rhs.count + numericCast(lhs.underestimateCount()))
```
Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: This fix LGTM.  Feel free to submit it as a separate PR and I'll merge immediately.

**Comment by gribozavr** [comment]
@nielsandriesse Thank you for your patch!  There are a few things in this PR:

1. A new overload for `+`.  This is a new API that needs to be proposed on swift-evolution.  To back your proposal, I would recommend writing tests that check the result type of adding all combinations between `{ range replaceable collection, collection, sequence, array, set }` and verifying that the API behavior makes sense.

2. A bug fix for `reserveCapacity` call.  This can be merged immediately if you submitted it as a separate PR.

3. Fixes for reference-typed collections.  The issue is that Swift collections library hasn't been tested with reference-typed collections at all, as far as I know.  Fixing these two spots won't help with the general issue.  If you are interested in fixing the issue, I would recommend that you start by writing tests for them: write minimal collection implementations like `MinimalSequence` and `MinimalRandomAccessRangeReplaceableCollection`, and tests like `./test/1_stdlib/RangeReplaceable.swift.gyb` that call into existing test cases.  This will allow you to quickly discover the most prominent issues, and propose a general strategy.

For now, I would recommend removing the fixes for reference-typed collections until we have a general story.

File: stdlib/public/core/RangeReplaceableCollectionType.swift (hunk 3 above)
Comment (nielsandriesse) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: Submitted as a separate PR: #736

File: stdlib/public/core/RangeReplaceableCollectionType.swift (hunk 2 above)
Comment (nielsandriesse) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: I’m in favor of removing both `CollectionType` overloads, like I mentioned in the PR description (I only added the new overload to match the existing one). I’ll submit the removal of these overloads (including tests) as a separate PR.

File: stdlib/public/core/RangeReplaceableCollectionType.swift (hunk 1 above)
Comment (PatrickPijnappel) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: We could just use `var result = C(lhs)`, as values type collections will generally be COW anyways so it would be equivalent to the plain assignment.

Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: No, that will typically cause a copy to be performed, even if it was redundant.

Comment (PatrickPijnappel) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: Why would such a copy be performed? It seems like a important optimization that is trivial to implement for any type already implementing COW, or am I missing something?

Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: It would be easy to implement, but not required to implement.  The default implementation does not do it (and can't, given the current language features).

Comment (PatrickPijnappel) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: Yeah in that sense, true, an arbitrary type conforming to `RangeReplaceableCollectionType` doesn't necessarily perform that optimization. Though, I'd still argue calling `C(lhs)` is the way to go:
- All stdlib collection types do perform this optimization, and many custom collection types are built on top of them so 'inherit' the optimization.
- The behavior does not become semantically *wrong* if a custom type doesn't implement it.
- That initializer performs a semantic copy, which is the behavior we need in the general case, because semantically, we're returning a new collection.
Or do you have a better alternative in mind? Because the current behavior is just plain incorrect for reference types, which I'd argue is much worse.

Comment (gribozavr) [inline stdlib/public/core/RangeReplaceableCollectionType.swift]: The library has to be not just correct, it has to be efficient.  Given that we know that we have other issues with reference-typed collections, I'm inclined to defer fixing this issue until we have tests that show that reference-typed collections work reasonably well in general.

I'm pretty sure that a lot of other algorithms rely on collections being value types, it is just that this particular piece of code happens to have a FIXME.  Once we know what the general issues are, we can start addressing them in a principled way, possibly adjusting the collections API along the way, to make it less error-prone to handle reference typed collections.

**Comment by CodaFi** [comment]
Hello @nielsandriesse.  As it's been quite some time since this pull request has been updated, and parts of it have already been merged (#736), I'm going to close this.  Please reopen it if you wish to continue working on the rest of the things Dmitri has requested.
